runtimes/protocol/notification.ts

```diff
@@ -0,0 +1,69 @@
+import { MessageType, ProtocolNotificationType } from './lsp'
+
+export interface NotificationContent {
+    readonly text: string
+    readonly title?: string
+}
+
+export namespace FollowupNotificationActionType {
+    export const Acknowledge = 'Acknowledge'
+}
+
+export type FollowupNotificationActionType = typeof FollowupNotificationActionType.Acknowledge
+
+export namespace NotificationActionType {
+    export const Url = 'Url'
+    export const Marketplace = 'Marketplace'
+}
+
+export type NotificationActionType =
+    | typeof NotificationActionType.Url
+    | typeof NotificationActionType.Marketplace
+    | typeof FollowupNotificationActionType.Acknowledge
+
+export interface NotificationAction {
+    readonly text: string
+    readonly type: NotificationActionType
+}
+
+export interface UrlAction extends NotificationAction {
+    readonly type: typeof NotificationActionType.Url
+    readonly url: string
+}
+
+export interface MarketplaceAction extends NotificationAction {
+    readonly type: typeof NotificationActionType.Marketplace
+}
+
+export interface AcknowledgeRequestAction extends NotificationAction {
+    readonly type: typeof FollowupNotificationActionType.Acknowledge
+}
+
+export interface NotificationParams {
+    readonly id?: string
```

Comment:
Do we really want to leave ID as optional?  For follow-up purposes and any other reason to perform an action/update against a notification received, we'd need an ID, correct?  If notifications don't have IDs in their source, it's trivial for a server to just generate a UUID or hash and just ignore it if not needed.  The concern is it seems reasonable that we will at some point have functionality that requires and ID and it would be better for servers to implement that as a requirement upfront.

Reply:
`Id` is currently needed for Toolkit notifications only. If there is followup expected, `id` would be needed. However, followup is not a common case. I would prefer to leave this optional.